cirq/protocols/apply_mixture_protocol.py

```diff
@@ -0,0 +1,309 @@
+    if left_shape != right_shape:
+        raise ValueError('Invalid target_tensor shape or selected axes. '
+                         'The selected left and right shape of target_tensor '
+                         'are not equal. Got {!r} and {!r}.'.format(
+                             left_shape, right_shape))
+    if val_qid_shape != left_shape:
+        raise ValueError('Invalid mixture qid shape is not equal to the '
+                         'selected left and right shape of target_tensor. '
+                         'Got {!r} but expected {!r}.'.format(
+                             val_qid_shape, left_shape))
+
+    # Check if the specialized method is present.
+    func = getattr(val, '_apply_mixture_', None)
+    if func is not None:
+        result = func(args)
+        if result is not NotImplemented and result is not None:
+
+            def err_str(buf_num_str):
+                return ("Object of type '{}' returned a result object equal to "
+                        "auxiliary_buffer{}. This type violates the contract "
+                        "that appears in apply_mixture's documentation.".format(
+                            type(val), buf_num_str))
+
+            assert result is not args.auxiliary_buffer0, err_str('0')
+            assert result is not args.auxiliary_buffer1, err_str('1')
+            return result
+
+    # Possibly use `apply_unitary`.
+    result = _apply_unitary(val, args)
+    if result is not None:
+        return result
+
+    # Fallback to using the object's `_mixture_` matrices.
+    prob_mix = mixture(val, None)
+    if prob_mix is not None:
+        return _apply_mixture(prob_mix, args)
+
+    # Don't know how to apply mixture. Fallback to specified default behavior.
+    if default is not RaiseTypeErrorIfNotProvided:
+        return default
+    raise TypeError(
+        "object of type '{}' has no _apply_mixture_, _apply_unitary_, "
+        "_unitary_, or _mixture_ methods (or they returned None or "
+        "NotImplemented).".format(type(val)))
+
+
+def _apply_unitary(val: Any, args: 'ApplyMixtureArgs') -> Optional[np.ndarray]:
+    """Attempt to use `apply_unitary` and return the result.
+
+    If `val` does not support `apply_unitary` returns None.
+    """
+    left_args = ApplyUnitaryArgs(target_tensor=args.target_tensor,
+                                 available_buffer=args.auxiliary_buffer0,
+                                 axes=args.left_axes)
+    left_result = apply_unitary(val, left_args, None)
+    if left_result is None:
+        return None
+    right_args = ApplyUnitaryArgs(target_tensor=np.conjugate(left_result),
+                                  available_buffer=args.auxiliary_buffer0,
+                                  axes=args.right_axes)
+    right_result = apply_unitary(val, right_args)
+    np.conjugate(right_result, out=right_result)
+    return right_result
+
+
+def _apply_unitary_fallback(
```

Comment:
The naming here is confusing. What's the actual conceptual distinction between apply unitary and apply unitary fallback?

Reply:
changed